parquet/src/main/java/org/apache/iceberg/parquet/BasePageIterator.java

```diff
@@ -0,0 +1,202 @@
+    this.triplesCount = initPage.getValueCount();
+    ValuesReader dlReader = null;
+    if (supportsVectorizedReads()) {
+      this.vectorizedDefinitionLevelReader = newVectorizedDefinitionLevelReader(desc);
+    } else {
+      dlReader = initPage.getDlEncoding().getValuesReader(desc, ValuesType.DEFINITION_LEVEL);
+      this.definitionLevels = newNonVectorizedDefinitionLevelReader(dlReader);
+    }
+    ValuesReader rlReader = initPage.getRlEncoding().getValuesReader(desc, ValuesType.REPETITION_LEVEL);
+    this.repetitionLevels = new PageIterator.ValuesReaderIntIterator(rlReader);
+    try {
+      BytesInput bytes = initPage.getBytes();
+      LOG.debug("page size {} bytes and {} records", bytes.size(), triplesCount);
+      LOG.debug("reading repetition levels at 0");
+      ByteBufferInputStream in = bytes.toInputStream();
+      rlReader.initFromPage(triplesCount, in);
+      LOG.debug("reading definition levels at {}", in.position());
+      if (supportsVectorizedReads()) {
+        this.vectorizedDefinitionLevelReader.initFromPage(triplesCount, in);
+      } else {
+        dlReader.initFromPage(triplesCount, in);
+      }
+      LOG.debug("reading data at {}", in.position());
+      initDataReader(initPage.getValueEncoding(), in, initPage.getValueCount());
+    } catch (IOException e) {
+      throw new ParquetDecodingException("could not read page " + initPage + " in col " + desc, e);
+    }
+  }
+
+  protected void initFromPage(DataPageV2 initPage) {
+    this.triplesCount = initPage.getValueCount();
+    this.repetitionLevels = newRLEIterator(desc.getMaxRepetitionLevel(), initPage.getRepetitionLevels());
+    if (supportsVectorizedReads()) {
+      this.vectorizedDefinitionLevelReader = newVectorizedDefinitionLevelReader(desc);
+    } else {
+      this.definitionLevels = newRLEIterator(desc.getMaxDefinitionLevel(), initPage.getDefinitionLevels());
+    }
+    LOG.debug("page data size {} bytes and {} records", initPage.getData().size(), triplesCount);
+    try {
+      initDataReader(initPage.getDataEncoding(), initPage.getData().toInputStream(), triplesCount);
+    } catch (IOException e) {
+      throw new ParquetDecodingException("could not read page " + initPage + " in col " + desc, e);
+    }
+  }
+
+  private IntIterator newRLEIterator(int maxLevel, BytesInput bytes) {
+    try {
+      if (maxLevel == 0) {
+        return new NullIntIterator();
+      }
+      return new RLEIntIterator(
+          new RunLengthBitPackingHybridDecoder(
+              BytesUtils.getWidthFromMaxInt(maxLevel),
+              bytes.toInputStream()));
+    } catch (IOException e) {
+      throw new ParquetDecodingException("could not read levels in page for col " + desc, e);
+    }
+  }
+
+  public void setDictionary(Dictionary dict) {
+    this.dictionary = dict;
+  }
+
+  protected abstract static class IntIterator {
+    abstract int nextInt();
+  }
+
+  static class ValuesReaderIntIterator extends IntIterator {
```

Comment:
This would work with a vectorized `ValuesReader` right?

Reply:
`ValuesReaderIntIterator` is only used in the non vectorized version `PageIterator`

Reply:
Maybe move this into the non-vectorized version then?

Reply:
@samarthjain, let's move these if they aren't used by the vectorized path back into the row-based path.

Reply:
I misspoke earlier. The `ValuesReaderIntIterator` is used in BasePageIterator as well.
`
    this.repetitionLevels = new ValuesReaderIntIterator(rlReader);
`
which is used by both the code paths.